waftools/generators/sources.py

```diff
@@ -21,20 +20,25 @@ def __file2string__(ctx, **kwargs):
         **kwargs
     )
 
-def __matroska_header__(ctx, **kwargs):
-    ctx(
-        rule   = __matroska_cmd__(ctx, '--generate-header'),
-        before = ("c",),
-        name   = os.path.basename(kwargs['target']),
-        **kwargs
-    )
-
-def __matroska_definitions__(ctx, **kwargs):
-    ctx(
-        rule   = __matroska_cmd__(ctx, '--generate-definitions'),
-        before = ("c",),
-        **kwargs
-    )
+@TaskGen.feature('ebml_header')
+def ebml_header(self):
+    setattr(self, 'rule', ' ') # waf doesn't print the task with no rule
+    target = getattr(self, 'target', None)
+    out = self.path.find_or_declare(target)
+    tmp = StringIO()
+    generate_C_header(tmp)
+    out.write(tmp.getvalue())
+    tmp.close()
+
+@TaskGen.feature('ebml_definitions')
+def ebml_definitions(self):
+    setattr(self, 'rule', ' ')
+    target = getattr(self, 'target', None)
+    out = self.path.find_or_declare(target)
+    tmp = StringIO()
+    generate_C_definitions(tmp)
+    out.write(tmp.getvalue())
+    tmp.close()
```

Comment:
Would it be possible to factorize some of this code? it's repeated 3 times with file2string.